waypoint/services/nats_service.py

```diff
@@ -27,42 +29,53 @@ def __init__(self, jetstream: JetStreamContext):
     async def _subscribe(
         self, *, group_id: str, wallet_id: str, topic: str, state: str, look_back: int
     ) -> JetStreamContext.PullSubscription:
-        try:
-            logger.trace(
-                "Subscribing to JetStream for wallet_id: {}, group_id: {}",
-                wallet_id,
-                group_id,
-            )
-            group_id = group_id or "*"
-            subscribe_kwargs = {
-                "subject": f"{NATS_STATE_SUBJECT}.{group_id}.{wallet_id}.{topic}.{state}",
-                "stream": NATS_STATE_STREAM,
-            }
 
-            # Get the current time in UTC
-            current_time = datetime.now(timezone.utc)
+        logger.trace(
+            "Subscribing to JetStream for wallet_id: {}, group_id: {}",
+            wallet_id,
+            group_id,
+        )
 
-            # Subtract look_back time from the current time
-            look_back_time = current_time - timedelta(seconds=look_back)
+        group_id = group_id or "*"
+        subscribe_kwargs = {
+            "subject": f"{NATS_STATE_SUBJECT}.{group_id}.{wallet_id}.{topic}.{state}",
+            "stream": NATS_STATE_STREAM,
+        }
 
-            # Format the time in the required format
-            start_time = look_back_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
-            config = ConsumerConfig(
-                deliver_policy=DeliverPolicy.BY_START_TIME,
-                opt_start_time=start_time,
-            )
-            subscription = await self.js_context.pull_subscribe(
-                config=config, **subscribe_kwargs
-            )
+        # Get the current time in UTC
+        current_time = datetime.now(timezone.utc)
 
-            return subscription
+        # Subtract look_back time from the current time
+        look_back_time = current_time - timedelta(seconds=look_back)
 
-        except BadSubscriptionError as e:
-            logger.error("BadSubscriptionError subscribing to NATS: {}", e)
-            raise
-        except Error as e:
-            logger.error("Error subscribing to NATS: {}", e)
-            raise
+        # Format the time in the required format
+        start_time = look_back_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
+        config = ConsumerConfig(
+            deliver_policy=DeliverPolicy.BY_START_TIME,
+            opt_start_time=start_time,
+        )
+
+        @retry(TimeoutError, delay=1, backoff=2, max_delay=16, logger=logger)
```

Comment:
I first recommended `retry`, but let's look into using `tenacity` instead. `retry` should work, but is unmaintained for many years now.

`tenacity` example a la GPT:
```py
from tenacity import retry, stop_after_attempt, wait_exponential

@retry(stop=stop_after_attempt(5), wait=wait_exponential(multiplier=1, min=1, max=10))
def unstable_function():
    print("Trying...")
    raise Exception("Failure")

unstable_function()
```

Reply:
See: https://github.com/jd/tenacity